### ISSUE TITLE
A few doubts about ccxt.pro and its roadmap

### ISSUE THREAD

**Issue by e36-cicd** [ccxt/ccxt#8074]
Hi,

We bought a short-term subscription in order to evaluate ccxt.pro. Before deciding whether to extend our subscription or not, we'd like to have a few things sorted out, if possible.

# Development Roadmap
Is it possible we see a quarterly roadmap with proper milestones for development and being able track it? From browsing some of the older issues, I see some promises on Q2 support for some exchanges, which haven't really been fulfilled yet (e.g. https://github.com/ccxt/ccxt/issues/6901).

This makes us worried about the further development of ccxt.pro and the ability to keep up with the ever-changing world of cryptocurrency exchanges...

# Exchange coverage
We'd like to be able to connect to the top exchanges for both spot and derivatives, but we found out some major exchanges are missing, for example: HuobiDM, Bybit, Bitflyer and Deribit.

The rule of thumb we follow is coingecko.com -> sort by normalized (for spot), or regular (for derivatives) 24h volume -> exchanges with Trust Score >= 9, and more than 300M$ of daily volume.

What is your take on this? When can we expect these to be added?

# Data normalization
Technically, we expected ccxt.pro to be a normalized data stream, that is, it will iron out all the different exchanges' quirks and not require us to be intimate with their requirements. For example, we didn't expect that for Bitmex, on each level of the book there are 3 entries (price, size, and id), instead of just (price, size) for most other exchanges.

There are a few other examples that surprised us, for example Binance futures - why are perpetual swaps for BTC/USDT are marked as 'future' instead of 'swap' like for Bitmex and others? We thought that 'future' is for dated, currency futures with expiration. 

Also, these futures (or options) that expire don't have any normalized 'expiration' date (or strike price) metadata that we can use, and we need to take a look at the 'info' member, which is exchange specific, just to derive that information - obviously this changes per exchange...

What is your view on this? Do you see it the same as us?

Thanks!

**Comment by kroitor** [comment]
Hi @e36-cicd !

On the development roadmap:

> Is it possible we see a quarterly roadmap with proper milestones for development and being able track it? From browsing some of the older issues, I see some promises on Q2 support for some exchanges, which haven't really been fulfilled yet (e.g. #6901).

The Q2 roadmap includes the futures support and positional trading, this is something we are still working on. In fact some results are already on the public CCXT repo, and we are going to upload more very soon. That will hopefully get us closer to the unification goals.

-------------------------------------------------------

On the exchange coverage:

> We'd like to be able to connect to the top exchanges for both spot and derivatives, but we found out some major exchanges are missing, for example: HuobiDM, Bybit, Bitflyer and Deribit.
> What is your take on this? When can we expect these to be added?

We are working on HuobiDM, Bybit and Deribit, we will add one or two of those three exchanges before the end of the year. Bitflyer is also on our plans for early 2021.

-------------------------------------------------------

On the unification: 

> For example, we didn't expect that for Bitmex, on each level of the book there are 3 entries (price, size, and id), instead of just (price, size) for most other exchanges.

BitMEX is an l3 orderbook this is the way they serve their orderbook. It's basically a one-liner (not an issue)in any language to drop the third extra value. We will consider adding an option to control that.

> There are a few other examples that surprised us, for example Binance futures - why are perpetual swaps for BTC/USDT are marked as 'future' instead of 'swap' like for Bitmex and others? We thought that 'future' is for dated, currency futures with expiration.
> Also, these futures (or options) that expire don't have any normalized 'expiration' date (or strike price) metadata that we can use, and we need to take a look at the 'info' member, which is exchange specific, just to derive that information - obviously this changes per exchange...

This is because you think of that functionality, as if it's a settled thing. In fact, it's been added a few months ago and it's still changing significantly. In other words, we are talking of the exchanges' APIs that are being developed + CCXT lib that is also being developed. In particular, the Binance's unusual terminology can be easily explained by historical reasons. First, Binance had just the spot markets and the margin trading mode. Then they added a `fapi`, standing for the _"futures API"_. Despite the name, all markets served with the `fapi` were perpetual futures contracts (aka swaps), not classic futures with delivery dates. Then they added the `dapi`, standing for the _"delivery API"_. Despite the name, the dapi served both the classic futures with delivery dates and the perpetual swaps. So, there's no systematic approach to these terms on the Binance side.  Later, Binance renamed those sections of the API, now they're mapped as follows:

- Binance `fapi` = future API = USDT-margined contracts (perpetual swaps only, but we have to call them "futures" for historical reason to keep the code working)
- Binance `dapi` = delivery API = COIN-margined contracts (both the futures with delivery dates and perpetual swaps, but we have to call them "delivery" for historical reasons to keep the code working)

We keep their API terms, since the implementation relies on those specific words for backward-compatibility. The plan is to rewrite this specific aspect of Binance implementation, however, there are certain other technical difficulties with Binance's markets classifications. Basically, it requires a unified symbol schema for swaps and delivery-futures, which is also on that Q2 roadmap, we've made some progress towards that during this year and we hope to finally roll it out soon.

> What is your view on this? Do you see it the same as us?

In short – we totally agree, this is not very intuitive, but we are aware of the issues and we are working hard towards the better unification.

----------------------------

Let us know if the above does not answer your questions.
We appreciate your feedback, and will of course address all that in the coming versions.
Feel free to reopen this issue or just ask further questions if any. 

Thank you for using CCXT Pro!

**Comment by e36-cicd** [comment]
Hi,

Thanks for the elaborate answer. A couple of follow-ups:

1. Regarding the Q2 roadmap - the issue I referred to was the roadmap for Q2 2020, and since it has already ended in 1st of July, I was wondering why hasn't HuobiDM been accomplished vs. what was promised in that issue.
2. Where can I see the exact list of Quarter-by-Quarter (or at least Q4 2020 + Q1 2021) issues that are to be addressed, in order to keep track of the progress being made? The best I could find was https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#pending-tasks but it's really lacking any visibility - is the task currently in progress, when is it expected to finish, and was it indeed finished...
3. Will it be possible to make a flag for ccxt.pro to choose which symbol 'versioning' you would like? As such, one could always use the older scheme (default), where if some major changes like Binance Futures+Delivery were added, a new version could make the support more unified.
4. May I add to the list of pending task the option to receive the differences in order book changes instead of snapshots? This will allow us to process the minimum required in order to act, rather than go through the entire book and find the diffs ourselves...

Just on a side note, appreciate the work you've done. I'm very well aware this is not trivial and intuitive, but that's exactly the added value ccxt.pro brings to the table, so obviously you need to make sure it's done right.

Thanks again!

**Comment by kroitor** [comment]
Hi!

> 1. Regarding the Q2 roadmap - the issue I referred to was the roadmap for Q2 2020, and since it has already ended in 1st of July, I was wondering why hasn't HuobiDM been accomplished vs. what was promised in that issue.

It has not been integrated yet, since there was a lot of other work, fixes and upgrades to the already integrated exchanges, maintenance, support and other tasks. We are aware of the demand and we hope to get it added asap.

> Where can I see the exact list of Quarter-by-Quarter (or at least Q4 2020 + Q1 2021) issues that are to be addressed, in order to keep track of the progress being made? The best I could find was https://github.com/ccxt/ccxt/blob/master/CONTRIBUTING.md#pending-tasks but it's really lacking any visibility - is the task currently in progress, when is it expected to finish, and was it indeed finished...

Check the commit log and the open/closed issues on GitHub. Apart from that we don't have a tracker yet. We will consider adding something like that.

> Will it be possible to make a flag for ccxt.pro to choose which symbol 'versioning' you would like? As such, one could always use the older scheme (default), where if some major changes like Binance Futures+Delivery were added, a new version could make the support more unified.

We hope to make it seamless to a certain extent, and yes, we will add the controls and flags where necessary.

> May I add to the list of pending task the option to receive the differences in order book changes instead of snapshots? This will allow us to process the minimum required in order to act, rather than go through the entire book and find the diffs ourselves...

That is already on our list, working on that specific aspect right now. Will let you know when it arrives.

Thank you for your feedback, we really appreciate it, and we will do our best to deliver more functionality as soon as we can!
